### PR TITLE
Bot api 3.2

### LINKED ISSUE #730: Add sticker management
Added new methods for working with stickers: getStickerSet, uploadStickerFile, createNewStickerSet, addStickerToSet, setStickerPositionInSet, and deleteStickerFromSet.
    Added the fields set_name and mask_position to the Sticker object, plus two new objects, StickerSet, and MaskPosition.

### PR DESCRIPTION
**Pull request by bomjacob**
Closes #730 

Note the lack of proper tests for StickerSet... it's hard test creation and such when we can't delete them afterwards.

Every bot method is documented as per new standard in #728. I have not added or changed the object docs for Sticker, StickerSet and MaskPosition however since that would potentially conflict with #728, so it's easier to just leave that to @Eldinnie :)

### CODE REVIEW

File: telegram/files/sticker.py [hunk 1]
```diff
@@ -72,5 +83,67 @@ def de_json(data, bot):
...
+        data['mask_position'] = MaskPosition.de_json(data.get('mask_position'), bot)
 
         return Sticker(**data)
+
+    @staticmethod
+    def de_list(data, bot):
+        if not data:
+            return list()
+
+        stickers = list()
```
Comment (thodnev) [inline telegram/files/sticker.py]: It's better to use comprehensions etc here like:
stickers = [Sticker.de_json(ix, bot) for ix in data]

File: telegram/files/sticker.py [hunk 2]
```diff
@@ -72,5 +83,67 @@ def de_json(data, bot):
...
+        self.name = name
+        self.title = title
+        self.contains_masks = contains_masks
+        self.stickers = stickers
+
+        self._id_attrs = (self.name,)
+
+    @staticmethod
+    def de_json(data, bot):
+        if not data:
```
Comment (thodnev) [inline telegram/files/sticker.py]: Probably we need to raise an exception here or make it return empty stickerset. Exception looks better in my opinion

File: telegram/files/sticker.py [hunk 3]
```diff
@@ -72,5 +83,67 @@ def de_json(data, bot):
...
+        return data
+
+
+class MaskPosition(TelegramObject):
+    FOREHEAD = 'forehead'
+    EYES = 'eyes'
+    MOUTH = 'mouth'
+    CHIN = 'chin'
+
+    def __init__(self, point, x_shift, y_shift, zoom, **kwargs):
```
Comment (thodnev) [inline telegram/files/sticker.py]: If it's used mainly for storage it's better to use __slots__

File: telegram/files/sticker.py (hunk 1 above)
Comment (bomjacob) [inline telegram/files/sticker.py]: Hmm, you're probably right, I just wrote it like we had in other places. I'll change it though.

File: telegram/files/sticker.py (hunk 2 above)
Comment (bomjacob) [inline telegram/files/sticker.py]: Nah, it's never been needed before in any of the other de_json, so we should be 100% fine

File: telegram/files/sticker.py (hunk 3 above)
Comment (bomjacob) [inline telegram/files/sticker.py]: Hmm. We don't use slots anywhere else, should we?

File: telegram/bot.py [hunk 4]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+        Args:
+            name (:obj:`str`): Short name of the sticker set that is used in t.me/addstickers/
+                URLs (e.g., animals)
+            timeout (:obj:`int` | :obj:`float`, optional): If this value is specified, use it as
+                the read timeout from the server (instead of the one specified during
+                creation of the connection pool).
+            **kwargs (:obj:`dict`): Arbitrary keyword arguments.
+
+        Returns:
+            :class:`telegramStickerSet`
```
Comment (Eldinnie) [inline telegram/bot.py]: Missing a `.` here (`telegram.StickerSet`)

File: telegram/bot.py [hunk 5]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+        data = {'name': name}
+
+        result = self._request.post(url, data, timeout=timeout)
+
+        return StickerSet.de_json(result, self)
+
+    def upload_sticker_file(self, user_id, png_sticker, timeout=None, **kwargs):
+        """
+        Use this method to upload a .png file with a sticker for later use in
+        :attr:`create_new_sticker_set` and :attr:`add_sticker_to_set methods` (can be used multiple
```
Comment (Eldinnie) [inline telegram/bot.py]: closing ``` is after methods, should close at _set

File: telegram/bot.py [hunk 6]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+        times).
+
+        Note:
+            The png_sticker argument can be either a file_id, an URL or a file from disk
+            ``open(filename, 'rb')``
+
+        Args:
+            user_id (int): User identifier of sticker file owner.
```
Comment (Eldinnie) [inline telegram/bot.py]: :obj:`int`

File: telegram/bot.py [hunk 7]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+        Returns:
+            :class:`telegramStickerSet`
+
+        Raises:
+            :class:`telegram.TelegramError`
+        """
+
+        url = '{0}/getStickerSet'.format(self.base_url)
+
+        data = {'name': name}
```
Comment (Eldinnie) [inline telegram/bot.py]: Definging `**kwargs` in arguments but not adding them to data is useless. I suggest removing them from arguments.

File: telegram/bot.py [hunk 8]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+        Returns:
+            :class:`telegram.File`: The uploaded File
+
+        Raises:
+            :class:`telegram.TelegramError`
+        """
+
+        url = '{0}/uploadStickerFile'.format(self.base_url)
+
+        data = {'user_id': user_id, 'png_sticker': png_sticker}
```
Comment (Eldinnie) [inline telegram/bot.py]: same note about `**kwargs`.

File: telegram/bot.py [hunk 9]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+            ``open(filename, 'rb')``
+
+        Args:
+            user_id (:obj:`int`): User identifier of created sticker set owner.
+            name (:obj:`str`): Short name of sticker set, to be used in t.me/addstickers/ URLs
+                (e.g., animals). Can contain only english letters, digits and underscores.
+                Must begin with a letter, can't contain consecutive underscores and
+                must end in "_by_<bot username>". <bot_username> is case insensitive.
+                1-64 characters.
+            title (:obj:`str`): Sticker set title, 1-64 characters
```
Comment (Eldinnie) [inline telegram/bot.py]: `.` at the end of the line

File: telegram/bot.py [hunk 10]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+            png_sticker (:obj:`str` | `filelike object`): Png image with the sticker, must be up
+                to 512 kilobytes in size, dimensions must not exceed 512px,
+                and either width or height must be exactly 512px. Pass a file_id as a String to
+                send a file that already exists on the Telegram servers, pass an HTTP URL as a
+                String for Telegram to get a file from the Internet, or upload a new one
+                using multipart/form-data.
+            emojis (:obj:`str`): One or more emoji corresponding to the sticker.
+            is_masks (:obj:`bool`, optional): Pass True, if a set of mask stickers should be
+                created.
+            mask_position (:class:`MaskPosition`, optional): Position where the mask should be
```
Comment (Eldinnie) [inline telegram/bot.py]: telegram.MaskPosition

File: telegram/bot.py [hunk 11]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+                placed on faces.
+            timeout (:obj:`int` | :obj:`float`, optional): If this value is specified, use it as
+                the read timeout from the server (instead of the one specified during
+                creation of the connection pool).
+            **kwargs (:obj:`dict`): Arbitrary keyword arguments.
```
Comment (Eldinnie) [inline telegram/bot.py]: again the `**kwargs`

File: telegram/bot.py [hunk 12]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+
+        Returns:
+            :obj:`bool` On success, ``True`` is returned.
```
Comment (Eldinnie) [inline telegram/bot.py]: missing colon (:obj:`bool`:)

File: telegram/bot.py [hunk 13]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+        """
+        Use this method to add a new sticker to a set created by the bot.
+
+        Note:
+            The png_sticker argument can be either a file_id, an URL or a file from disk
+            ``open(filename, 'rb')``
+
+        Args:
+            user_id (:obj:`int`): User identifier of created sticker set owner.
+            name (:obj:`str`): Sticker set name
```
Comment (Eldinnie) [inline telegram/bot.py]: `.` at end of line

File: telegram/bot.py [hunk 14]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+            png_sticker (:obj:`str` | `filelike object`): Png image with the sticker, must be up
+                to 512 kilobytes in size, dimensions must not exceed 512px,
+                and either width or height must be exactly 512px. Pass a file_id as a String to
+                send a file that already exists on the Telegram servers, pass an HTTP URL as a
+                String for Telegram to get a file from the Internet, or upload a new one
+                using multipart/form-data.
+            emojis (:obj:`str`): One or more emoji corresponding to the sticker.
+            mask_position (:class:`MaskPosition`, optional): Position where the mask should be
```
Comment (Eldinnie) [inline telegram/bot.py]: telegram.MaskPosition

File: telegram/bot.py [hunk 15]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+                placed on faces.
+            timeout (:obj:`int` | :obj:`float`, optional): If this value is specified, use it as
+                the read timeout from the server (instead of the one specified during
+                creation of the connection pool).
+            **kwargs (:obj:`dict`): Arbitrary keyword arguments.
+
+        Returns:
+            :obj:`bool` On success, ``True`` is returned.
```
Comment (Eldinnie) [inline telegram/bot.py]: colon

File: telegram/bot.py [hunk 16]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+                send a file that already exists on the Telegram servers, pass an HTTP URL as a
+                String for Telegram to get a file from the Internet, or upload a new one
+                using multipart/form-data.
+            emojis (:obj:`str`): One or more emoji corresponding to the sticker.
+            mask_position (:class:`MaskPosition`, optional): Position where the mask should be
+                placed on faces.
+            timeout (:obj:`int` | :obj:`float`, optional): If this value is specified, use it as
+                the read timeout from the server (instead of the one specified during
+                creation of the connection pool).
+            **kwargs (:obj:`dict`): Arbitrary keyword arguments.
```
Comment (Eldinnie) [inline telegram/bot.py]: **kwargs

File: telegram/bot.py [hunk 17]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+
+        return result
+
+    def set_sticker_position_in_set(self, sticker, position, timeout=None, **kwargs):
+        """
+        Use this method to move a sticker in a set created by the bot to a specific position.
+
+        Args:
+            sticker (:obj:`str`): File identifier of the sticker
+            position (:obj:`int`): New sticker position in the set, zero-based
```
Comment (Eldinnie) [inline telegram/bot.py]: dots on this line and above.

File: telegram/bot.py [hunk 18]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+            timeout (:obj:`int` | :obj:`float`, optional): If this value is specified, use it as
+                the read timeout from the server (instead of the one specified during
+                creation of the connection pool).
+            **kwargs (:obj:`dict`): Arbitrary keyword arguments.
```
Comment (Eldinnie) [inline telegram/bot.py]: **kwargs

File: telegram/bot.py [hunk 19]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+
+        Returns:
+            :obj:`bool` On success, ``True`` is returned.
```
Comment (Eldinnie) [inline telegram/bot.py]: colon

File: telegram/bot.py [hunk 20]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+        result = self._request.post(url, data, timeout=timeout)
+
+        return result
+
+    def delete_sticker_from_set(self, sticker, timeout=None, **kwargs):
+        """
+        Use this method to delete a sticker from a set created by the bot.
+
+        Args:
+            sticker (:obj:`str`): File identifier of the sticker
```
Comment (Eldinnie) [inline telegram/bot.py]: dot at end

File: telegram/bot.py [hunk 21]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+            timeout (:obj:`int` | :obj:`float`, optional): If this value is specified, use it as
+                the read timeout from the server (instead of the one specified during
+                creation of the connection pool).
+            **kwargs (:obj:`dict`): Arbitrary keyword arguments.
```
Comment (Eldinnie) [inline telegram/bot.py]: **kwargs

File: telegram/bot.py [hunk 22]
```diff
@@ -2369,6 +2369,218 @@ def unpin_chat_message(self, chat_id, timeout=None, **kwargs):
...
+
+        Returns:
+            :obj:`bool` On success, ``True`` is returned.
```
Comment (Eldinnie) [inline telegram/bot.py]: colon

File: telegram/files/sticker.py (hunk 3 above)
Comment (Eldinnie) [inline telegram/files/sticker.py]: constants all over the lib. Suggesting same approach here.

**Comment by bomjacob** [comment]
The failed build is due to a possible bug in telegrams documentation. I've already contacted botsupport. Should I add an overwrite so our builds pass or do we wanna wait for a reply?

File: telegram/files/sticker.py (hunk 2 above)
Comment (tsnoam) [inline telegram/files/sticker.py]: i agree with @bomjacob

File: telegram/bot.py (hunk 7 above)
Comment (tsnoam) [inline telegram/bot.py]: Good catch @Eldinnie 
`kwargs` are automatically passed when using the `@message` decorator, but if it's not being used, than they should be pushed in the `data` dictionary.
the reason for that, is to allow forward compatibility (lets say telegram adds some arguments t an existing API call, then users can use the new API even before we release an official support).

Comment (tsnoam) [inline telegram/bot.py]: anyway, we can fix that in another PR as it's not the only place where we missed that.

**Pull request merged**